log-manager/src/main/java/io/airlift/log/Logging.java

```diff
@@ -160,6 +160,12 @@ private static Level getEffectiveLevel(java.util.logging.Logger logger)
         return Level.fromJulLevel(level);
     }
 
+    public void clearLevel(String loggerName)
+    {
+        loggers.computeIfAbsent(loggerName, java.util.logging.Logger::getLogger)
```

Comment:
We may not need to hold on to the logger once the level is cleared

Reply:
So, instead this?

```
        java.util.logging.Logger logger = loggers.get(loggerName);
        if (logger == null) {
            logger = java.util.logging.Logger.getLogger(loggerName);
        }
        logger.setLevel(null);

```

Reply:
We could also clear the map but that's not really implied by the method contract.

Reply:
Yeah, that's better

Reply:
Providing clearLevel method suggests that calling `setLevel + clearLevel` (with varying names) is a supported thing.
So we should try to indeed support it, eg avoid memory leak. 
I suggest that in this method
```
java.util.logging.Logger logger = loggers.get(loggerName);
```
be replaced with
```
java.util.logging.Logger logger = loggers.remove(loggerName);
```

After that, there still is possibility that we have a logger in `loggers`, which has null level (inconsistency), so
I would certainly consider synchronizing `setLevel` and `clearLevel` methods to avoid this.
(then no `loggers` becomes a non-concurrent map)

Reply:
TBH - I'm not sure why `Logging` caches the logger in `setLevel()`. `java.util.logging.Logger.getLogger()` has its own cache internally. The logger cached in `loggers` would only be referenced again if the level is changed (or now cleared). But, I'm new to this code base. Maybe we should remove `loggers` altogether?

Reply:
I bisecting and see that the cache in `Logging` was added here:

```
commit 117e90fbe3ae7ef2443ab658e671197b7d0e7cfb
Author: David Phillips <david@acz.org>
Date:   Fri May 15 16:34:21 2015 -0700

    Fix log levels being reset by garbage collection
```

So, maybe @electrum can comment/review

Reply:
The JDK keeps a weak reference to the logger. That's why we have to keep one in this class.

Reply:
I don't think it's a big deal in practice, but let's remove it for completeness and consistency. It's unlikely anyone would be using these methods very "dynamically", or for large numbers of loggers, so any apparent leak would be generally bounded in size. In fact, that would happen anyway if they go around this class and clear the log levels on their own by calling the JDK methods directly.

Reply:
done

Reply:
> if they go around this class and clear the log levels on their own by calling the JDK methods directly

Which is what I'm having to do now BTW without this method.

Reply:
what about synchronization (https://github.com/airlift/airlift/pull/838#discussion_r414422707)?

- thread A calls setLevel, entering it into the map
- thread B calls clearLogger, removing it from the map
- thread B calls logger.setLevel(null)
- thread A calls logger.setLevel(DEBUG)

Since both calls are "concurrent", it's whatever whether the end state will be DEBUG or null.
But, the end state  is "DEBUG... until GC, then reset" -- that doesn't seem acceptable.

@martint any reason not to make the methods synchronized?

Reply:
A simpler solution is to leave it in the map as it was in the original version of this PR.

Reply:
ping @martint